src/Sanitizers.cpp

```diff
@@ -0,0 +1,70 @@
+#include "Sanitizers.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+
+namespace Halide {
+namespace Internal {
+
+namespace {
+
+Stmt call_extern_and_assert(const std::string& name, const std::vector<Expr>& args) {
+    Expr call = Call::make(Int(32), name, args, Call::Extern);
+    std::string call_result_name = unique_name(name + "_result");
+    Expr call_result_var = Variable::make(Int(32), call_result_name);
+    return LetStmt::make(call_result_name, call,
+                         AssertStmt::make(EQ::make(call_result_var, 0), call_result_var));
+}
+
+class InjectMSANHelpers : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const ProducerConsumer *op) {
+        Stmt produce, update, consume;
+
+        produce = mutate(op->produce);
+        if (op->update.defined()) {
+            update = mutate(op->update);
+        }
+        consume = mutate(op->consume);
+
+        Expr buffer = Variable::make(type_of<struct buffer_t *>(), op->name + ".buffer");
+        Stmt annotate = call_extern_and_assert("halide_msan_annotate_buffer_is_initialized", {buffer});
+
+        if (op->update.defined()) {
+            update = Block::make(update, annotate);
+        } else {
+            produce = Block::make(produce, annotate);
+        }
+
+        stmt = ProducerConsumer::make(op->name, produce, update, consume);
+    }
+
+    void visit(const Call *op) {
+        // convert all calls to halide_copy_to_host() into a pair of calls:
+        // halide_copy_to_host() + halide_msan_annotate_buffer_is_initialized()
+        if (op->call_type == Call::Extern && op->name == "halide_copy_to_host") {
```

Comment:
Yeah, this can has more worms than I expected. This would be the first case I'm aware of where a runtime module needs both "on" and "off" implementations. Options:

(1) Add an "msan-off" module with a stub to handle this case. (Yuck, ugly precedent as mentioned above.)
(2) Add a global function pointer for the "annotate" function (defaults to null) that will be filled in when msan is enabled that can be used by copy-to-host. Still yuck: avoids needing an msan-off module but adds a new weird special case.
(3) Punt on this corner case entirely and tell this user that they need to add an explicit call, wrapped in an #ifdef, which isn't entirely unreasonable since it's not uncommon to need explicit ANNOTATE_MEMORY() calls in spots for MSAN (i.e.: just add a comment saying "sorry, we know this is suboptimal")
(4) Punt on this PR entirely and revisit the idea of properly enabling the actual LLVM MSAN pass.